runtime/src/storage/tokio.rs

```diff
@@ -0,0 +1,300 @@
+use crate::Error;
+use commonware_utils::{from_hex, hex};
+use prometheus_client::metrics::{counter::Counter, gauge::Gauge};
+use prometheus_client::registry::Registry;
+use std::io::SeekFrom;
+use std::{path::PathBuf, sync::Arc};
+use tokio::{
+    fs,
+    io::{AsyncReadExt, AsyncSeekExt, AsyncWriteExt},
+    sync::Mutex as AsyncMutex,
+};
+
+pub struct Metrics {
+    pub open_blobs: Gauge,
+    pub storage_reads: Counter,
+    pub storage_read_bytes: Counter,
+    pub storage_writes: Counter,
+    pub storage_write_bytes: Counter,
+}
+
+impl Metrics {
+    /// Initialize the `Metrics` struct and register the metrics in the provided registry.
+    pub fn new(registry: &mut Registry) -> Self {
+        let metrics = Self {
+            open_blobs: Gauge::default(),
+            storage_reads: Counter::default(),
+            storage_read_bytes: Counter::default(),
+            storage_writes: Counter::default(),
+            storage_write_bytes: Counter::default(),
+        };
+
+        registry.register(
+            "open_blobs",
+            "Number of open blobs",
+            metrics.open_blobs.clone(),
+        );
+        registry.register(
+            "storage_reads",
+            "Total number of disk reads",
+            metrics.storage_reads.clone(),
+        );
+        registry.register(
+            "storage_read_bytes",
+            "Total amount of data read from disk",
+            metrics.storage_read_bytes.clone(),
+        );
+        registry.register(
+            "storage_writes",
+            "Total number of disk writes",
+            metrics.storage_writes.clone(),
+        );
+        registry.register(
+            "storage_write_bytes",
+            "Total amount of data written to disk",
+            metrics.storage_write_bytes.clone(),
+        );
+
+        metrics
+    }
+}
+
+#[derive(Clone)]
+pub struct Config {
+    pub storage_directory: PathBuf,
+}
+
+impl Config {
+    pub fn new(storage_directory: PathBuf) -> Self {
+        Self { storage_directory }
+    }
+}
+
+#[derive(Clone)]
+pub struct Storage {
+    metrics: Arc<Metrics>,
+    fs: Arc<AsyncMutex<()>>,
+    cfg: Config,
+}
+
+impl Storage {
+    pub fn new(metrics: &mut Registry, cfg: Config) -> Self {
+        Self {
+            metrics: Arc::new(Metrics::new(metrics)),
+            fs: AsyncMutex::new(()).into(),
+            cfg,
+        }
+    }
+}
+
+pub struct Blob {
+    metrics: Arc<Metrics>,
+    partition: String,
+    name: Vec<u8>,
+    file: Arc<AsyncMutex<(fs::File, u64)>>,
```

Comment:
Would you mind carrying over the comment that was here?
```

    // Files must be seeked prior to any read or write operation and are thus
    // not safe to concurrently interact with. If we switched to mapping files
    // we could remove this lock.
    //
    // We also track the virtual file size because metadata isn't updated until
    // the file is synced (not to mention it is a lot less fs calls).
   ```

Reply:
done